openfasoc/generators/gdsfactory-gen/process_input.py

```diff
@@ -0,0 +1,233 @@
+import nltk
+from relational import GlayoutCode, parse_direction
+import io
+from pathlib import Path
+import copy
+from typing import Optional, Union, Literal
+
+
+
+class Session:
+	"""The session stores all relevant information for producing code from a conversation"""
+	
+	generic_prompt = """Place a cell, move a cell, route, create a parameter, or define a variable.
+You can also dump code or save this conversation, or enter "help" to see supported syntax in detail
+What would you like to do?"""
+    
+	def __init__(self, outputstream: io.IOBase, inputstream: Optional[io.IOBase]=None, toplvlname: Optional[str]=None):
+		"""initialize a conversation and greet the user
+		Args:
+			outputstream (io.IOBase): used to print outputs
+			inputstream (io.IOBase): saved for (optionally) reading in user input, also just provide a string
+			NOTE: if input stream not provided, str input must be provided
+			toplvlname (str): name of the toplevel component. in string only mode, you can input toplvl name using this arg
+		"""
+		self.inputstream = inputstream
+		self.outputstream = outputstream
+		# greet the user and get a top level name for the component
+		if toplvlname is None:
+			if inputstream is None:
+				raise RuntimeError("you must specify AT LEAST one of inputstream or name")
+			self.print_to_stream("Hello!")
+			self.print_to_stream("Please provide a name for the Component you want to create")
+			self.print_to_stream("remember, this will be the name of your top level component: ")
+			self.name = self.read_from_stream().strip()
+		else:
+			self.name = str(toplvlname).strip()
+		# save a list of responses to recreate this component from a .conv file
+		self.conversation_responses = list()
+		self.conversation_responses.append(self.name)
+		# init the code object
+		self.code = GlayoutCode(self.name)
+		# create a backup that goes back exactly one call to process_next_input
+		self.backup = self.__backup()
+	
+
+	def __backup(self):
```

Comment:
It would be good to add docstrings for these functions as well.

Reply:
done